openquake/hazardlib/near_fault.py

```diff
@@ -0,0 +1,476 @@
+    s = np.array([sxx, sxy, sxz, sxy, syy, syz, sxz, syz,
+                 szz]).reshape(3, 3)
+
+    cx = np.sum(pa[:, 0].reshape(2, 1) * (nx ** 2 - 1) +
+                pa[:, 1].reshape(2, 1) * [nx * ny] +
+                pa[:, 2].reshape(2, 1) * (nx * nz))
+    cy = np.sum(pa[:, 0].reshape(2, 1) * [nx * ny] +
+                pa[:, 1].reshape(2, 1) * [ny ** 2 - 1] +
+                pa[:, 2].reshape(2, 1) * [ny * nz])
+    cz = np.sum(pa[:, 0].reshape(2, 1) * [nx * nz] +
+                pa[:, 1].reshape(2, 1) * [ny * nz] +
+                pa[:, 2].reshape(2, 1) * [nz ** 2 - 1])
+    c = np.array([cx, cy, cz]).reshape(3, 1)
+    p_intersect = np.linalg.solve(s, c)
+
+    vector1 = (p_intersect.flatten() - seg2_end) / \
+        sum((p_intersect.flatten() - seg2_end) ** 2) ** 0.5
+    vector2 = (seg2_start - seg2_end) / \
+        sum((seg2_start - seg2_end) ** 2) ** 0.5
+    vector3 = (np.array(seg1_end) - np.array(seg1_start)) / \
+        sum((np.array(seg1_end) - np.array(seg1_start)) ** 2) ** 0.5
+    vector4 = (p_intersect.flatten() - np.array(seg1_start)) / \
+        sum((p_intersect.flatten() - np.array(seg1_start)) ** 2) ** 0.5
+
+    return p_intersect, vector1, vector2, vector3, vector4
+
+
+def directp(node0, node1, node2, node3, hypocenter, reference, pp):
+    """
+    Get the Direct Point and the corresponding E-path desrcibed in: "Final
+    report of the NGA-West2 directivity working group." (2013) by Spudich,
+    Paul, et al. and publish in PEER report, page 85- 96. Also obtain the
+    flag if the DPP calculation keep going on the next fault segment(in the
+    case of multi-segment).
+    :param node0:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of one vertices on the target fault
+        segment.
+    :param node1:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of one vertices on the target fault
+        segment. Note, the order should be clockwise.
+    :param node2:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of one vertices on the target fault
+        segment. Note, the order should be clockwise.
+    :param node3:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of one vertices on the target fault
+        segment. Note, the order should be clockwise.
+    :param hypocenter:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of floating hypocenter on each segment
+        calculation. In the method, we take the direction point of the
+        previous fault patch as hypocentre for the current fault patch.
+    :param reference:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of reference point for projection
+    :param pp:
+        the projection point pp on the patch plane,
+        format in numpy array
```

Comment:
The directp method can only calculate for single fault with one patch, so there is only one patch is possible. Pp is obtained by one plane equation.
The description might be misleading.
I will change it to 
 :param pp:
+        the projection of the site onto the plane containing the fault slipped area
+        a numpy array
(that is the description in the PEER report)